src/app/libp2p_helper/src/libp2p_helper/main.go

```diff
@@ -446,12 +446,21 @@ func (s *subscribeMsg) run(app *app) (interface{}, error) {
 			return pubsub.ValidationIgnore
 		}
 
+		deadline, ok := ctx.Deadline()
+		if !ok {
+			app.P2p.Logger.Errorf("no deadline set on validation context")
+			defer app.ValidatorMutex.Unlock()
+			delete(app.Validators, seqno)
```

Comment:
Should this delete not precede the `Unlock`? It seems like the validator mutex ought to be held when we are modifying `app.Validators`. (Similarly in the `if` above, if so.)

Reply:
It's because of what `defer` does in Go. It's honestly a bizarre language construct (but it is kind of useful). It basically stages `app.ValidationMutex.Unlock()` so that it will be called before the function returns to the caller. This follows the pattern for locking and unlocking in other parts of the libp2p_helper code.